src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonArrayConverter.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    /// <summary>
+    /// Converter for <cref>System.Array</cref>.
+    /// </summary>
+    internal sealed class JsonArrayConverter<TCollection, TElement>
+        : JsonIEnumerableDefaultConverter<TCollection, TElement>
+        where TCollection: IEnumerable
+    {
+        internal override bool CanHaveIdMetadata => false;
+
+        protected override void Add(TElement value, ref ReadStack state)
+        {
+            ((List<TElement>)state.Current.ReturnValue!).Add(value);
```

Comment:
In these cases since creation of `state.Current.ReturnValue` is completely self-contained within the file via `CreateCollection()`, the Assert doesn't add much IMHO once the general pattern is noticed. However to clear the intent, I'll add the Assert (and another one below dealing with the enumerator).